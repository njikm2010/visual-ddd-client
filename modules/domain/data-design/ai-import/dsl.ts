import { DataObjectReferenceCardinality, DataObjectTypeName } from '../dsl/dsl';

export namespace AIImport {
  export interface DataObjectReferenceType {
    type: DataObjectTypeName.Reference;
    target: string;
    property: string;
    cardinality: DataObjectReferenceCardinality;
  }

  export interface DataObjectNormalType {
    type: Exclude<DataObjectTypeName, DataObjectTypeName.Reference>;
  }

  export interface DataObjectProperty {
    name: string;
    title: string;
    primaryKey?: boolean;
    type: DataObjectNormalType | DataObjectReferenceType;
  }

  export interface DataObject {
    name: string;
    title: string;
    properties: DataObjectProperty[];
  }
}
