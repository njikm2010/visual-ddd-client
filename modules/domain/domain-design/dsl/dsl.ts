export type UUID = string;

/**
 * 核心元数据
 */
export interface IDDSL {
  /**
   * 自动生成的唯一标识符
   */
  uuid: UUID;
}

/**
 * 关系描述
 */
export enum RelationShipDSL {
  /**
   * 对象关系
   * https://www.cnblogs.com/zhongj/p/11169780.html
   *
   * dependency 依赖关系, 在Java中，类A当中使用了类B，其中类B是作为类A的**方法参数**、方法中的局部变量、返回值、或者静态方法调用
   * association 关联关系、作为成员强依赖时
   * aggregation 聚合
   * composition 组合(统一生命周期的整体和部分的关系)
   */
  Dependency = 'dependency',
  Association = 'association',
  Aggregation = 'aggregation',
  Composition = 'composition',
}

/**
 * 引用表述
 */
export interface ReferenceDSL {
  referenceId: string;
  name?: string;
}

export interface MetaDSL {
  key?: string;
  value?: string;

  /**
   * 内部属性，不会在节点上显示
   */
  inner?: boolean;
}

export interface NameDSL extends IDDSL {
  /**
   * 唯一标识符, 用户输入
   */
  name: string;

  /**
   * 中文名称, 默认为 空字符串
   */
  title?: string;

  /**
   * 详细描述信息，默认为空字符串
   */
  description?: string;

  /**
   * 元数据，由开发者自行扩展,  DSL 本身不会关心其内容，DSL 消费者会关心
   */
  meta?: MetaDSL[];
}

/**
 * 标识符命名规则
 */
export type NameCase = 'CamelCase' | 'camelCase' | 'SNAKE_CASE' | 'snake_case';

export type BaseType =
  | 'String'
  | 'Integer'
  | 'Long'
  | 'Double'
  | 'Float'
  | 'Date'
  | 'Boolean'
  | 'BigDecimal'
  | 'Char'
  | 'Byte'
  | 'Short'
  | 'Void';

export const BaseTypeInArray: BaseType[] = [
  'String',
  'Integer',
  'Long',
  'Double',
  'Float',
  'Date',
  'Boolean',
  'BigDecimal',
  'Char',
  'Byte',
  'Short',
  'Void',
];

export type ContainerType = 'List' | 'Set' | 'Map';
export const ContainerTypeInArray: ContainerType[] = ['List', 'Set', 'Map'];

export enum TypeType {
  Base = 'base',
  Container = 'container',
  Reference = 'reference',
}

/**
 * 类型应用的表示
 * 基础类型: String, Integer, Long, Double, Float, Date, Boolean, BigDecimal, Char, Byte, Short, Void
 * 集合类型：List<T>, Set<T>, Map<K, V>
 * 引用类型：
 */
export type TypeDSL =
  | {
      // 基础类型
      type: TypeType.Base;
      name: BaseType;
    }
  | {
      // 容器数据
      type: TypeType.Container;
      name: ContainerType;

      /**
       * 泛型变量, 默认为 Void
       */
      params: { [key: string]: TypeDSL | undefined };
    }
  | {
      // 引用数据
      type: TypeType.Reference;
      /**
       * 引用 ID, 这个通常是不变的
       */
      referenceId: string;

      // 冗余字段，方便回显, 并不可靠
      name: string;
    };

/**
 * 访问控制
 */
export type AccessDSL = 'public' | 'private' | 'protected';

/**
 * 属性定义
 */
export interface PropertyDSL extends NameDSL {
  /**
   * 成员类型
   */
  type?: TypeDSL;

  /**
   * 访问控制, 默认 public
   */
  access?: AccessDSL;
}

/**
 * 方法参数定义
 */
export interface ParameterDSL extends NameDSL {
  /**
   * 参数
   */
  type?: TypeDSL;
}

/**
 * 方法定义
 */
export interface MethodDSL extends NameDSL {
  /**
   * 访问控制, 默认 public
   */
  access?: AccessDSL;

  /**
   * 是否为抽象方法, 默认 false
   */
  abstract?: boolean;

  /**
   * 方法参数
   */
  parameters: ParameterDSL[];

  /**
   * 方法返回值，当为空时相当于 void
   * return 是关键字，所以这里使用 result
   */
  result?: TypeDSL;
}

/**
 * 类定义
 */
export interface ClassDSL extends NameDSL {
  /**
   * 继承的类
   */
  extends?: string;

  /**
   * 扩展的接口
   */
  implements?: string[];

  /**
   * 是否为抽象类，默认为 false
   */
  abstract?: boolean;

  /**
   * 实例属性
   */
  properties: PropertyDSL[];

  /**
   * 实例方法
   */
  methods: MethodDSL[];

  /**
   * 类属性
   */
  classProperties: PropertyDSL[];

  /**
   * 类方法，静态方法，默认为 []
   */
  classMethods: MethodDSL[];
}

/**
 * 实体定义
 */
export interface EntityDSL extends ClassDSL {
  /**
   * 标记为聚合根
   */
  isAggregationRoot?: boolean;

  /**
   * 声明实体 id uuid
   */
  id: UUID;
}

/**
 * 值对象定义
 */
export interface ValueObjectDSL extends ClassDSL {}

export type EnumBaseType = 'string' | 'number';

export interface EnumMemberDSL extends NameDSL {
  /**
   * 后期进行转换
   */
  code: string;
}

/**
 * 枚举类型定义
 */
export interface EnumDSL extends NameDSL {
  baseType: EnumBaseType;
  members: EnumMemberDSL[];
}

/**
 * 命令来源
 */
export type SourceDSL = {
  http: {
    enabled: boolean;
  };
  rpc: {
    enabled: boolean;
  };
  /**
   * 外部事件
   */
  event: {
    enabled: boolean;
    /**
     * 事件名
     */
    value: string;
  };
  schedule: {
    enabled: boolean;
    /**
     *  cron 表达式
     */
    value: string;
  };
};

export type RepositoryDSL = 'create' | 'modify' | 'remove';

/**
 * 命令定义
 */
export interface CommandDSL extends NameDSL {
  /**
   * 命令来源
   */
  source: SourceDSL;

  /**
   * 绑定的仓储能力
   */
  repository: RepositoryDSL;

  /**
   * 事件是否是需要发送, 默认为 false
   */
  eventSendable: boolean;

  /**
   * 属性
   */
  properties: PropertyDSL[];

  /**
   * 命令属性
   */
  eventProperties: PropertyDSL[];

  /**
   * 所属的聚合
   */
  aggregation?: ReferenceDSL;

  /**
   * 返回值
   */
  result?: TypeDSL;
}

/**
 * 规则定义
 */
export interface RuleDSL extends NameDSL {
  /**
   * 聚合当前规则的父级
   */
  aggregator?: ReferenceDSL;
}

/**
 * 聚合定义
 */
export interface AggregationDSL extends NameDSL {
  // 实体、值对象等关联关系不在这里定义，Model 中已经包含了上下文信息
  /**
   * 颜色标记
   */
  color: string;
}

/**
 * DTO 定义
 */
export interface DTODSL extends ClassDSL {}

/**
 * 查询定义
 */
export interface QueryDSL extends NameDSL {
  /**
   * 查询来源
   */
  source: SourceDSL;

  /**
   * 属性
   */
  properties: PropertyDSL[];

  /**
   * 返回值
   */
  result?: TypeDSL;

  /**
   * 是否分页结果
   */
  pagination: boolean;
}
