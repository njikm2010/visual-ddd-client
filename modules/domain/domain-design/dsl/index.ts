export * from './dsl';
export * from './factory';
export * from './validators';
export * from './constants';
export * from './components';
export * from './stringify';
export * from './reactify';
export * from './dependency';
