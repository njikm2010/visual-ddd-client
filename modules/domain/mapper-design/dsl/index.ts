export * from './dsl';
export * from './factory';
export * from './constants';
