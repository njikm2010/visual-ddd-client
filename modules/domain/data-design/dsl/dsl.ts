import { NameDSL } from '@/modules/domain/domain-design/dsl/dsl';

/**
 * 类型
 */
export enum DataObjectTypeName {
  Boolean = 'Boolean',
  Date = 'Date',
  DateTime = 'DateTime',
  Timestamp = 'Timestamp',
  Integer = 'Integer',
  Decimal = 'Decimal',
  Long = 'Long',
  Double = 'Double',
  Float = 'Float',
  String = 'String',
  Text = 'Text',
  LongText = 'LongText',
  JSON = 'JSON',
  Reference = 'Reference',
}

/**
 * 布尔值
 */
export type DataObjectBoolean = {
  type: DataObjectTypeName.Boolean;
  defaultValue?: boolean;
};

/**
 * 日期
 */
export type DataObjectDate = {
  type: DataObjectTypeName.Date;
  defaultValue?: `${number}-${number}-${number}`;
};

/**
 * 日期时间
 */
export type DataObjectDateTime = {
  type: DataObjectTypeName.DateTime;
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 时间戳
 */
export type DataObjectTimestamp = {
  type: DataObjectTypeName.Timestamp;
  /**
   * 从 1970-01-01 00:00:00 开始的时间戳
   */
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 整数
 */
export type DataObjectInteger = {
  type: DataObjectTypeName.Integer;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * 小数
 */
export type DataObjectDecimal = {
  type: DataObjectTypeName.Decimal;
  defaultValue?: number;

  /**
   * 精度, 默认 10, 最大为 65
   */
  precision?: number;

  /**
   * 小数位, 默认 0
   */
  scale?: number;
};

/**
 * 长整型
 */
export type DataObjectLong = {
  type: DataObjectTypeName.Long;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Double
 */
export type DataObjectDouble = {
  type: DataObjectTypeName.Double;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Float
 */
export type DataObjectFloat = {
  type: DataObjectTypeName.Float;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * String
 */
export type DataObjectString = {
  type: DataObjectTypeName.String;
  defaultValue?: string;

  /**
   * 对应 varchar
   * 0 to 65,535.
   * 默认 255
   */
  length?: number;
};

/**
 * Text
 */
export type DataObjectText = {
  type: DataObjectTypeName.Text;
};

/**
 * 长文本
 */
export type DataObjectLongText = {
  type: DataObjectTypeName.LongText;
};

export enum DataObjectReferenceCardinality {
  OneToOne = 'OneToOne',
  OneToMany = 'OneToMany',
  ManyToOne = 'ManyToOne',
  ManyToMany = 'ManyToMany',
}

/**
 * 字段引用
 */
export type DataObjectReference = {
  type: DataObjectTypeName.Reference;

  /**
   * 目标对象 ID
   */
  target?: string;

  /**
   * 目标对象字段 ID
   */
  targetProperty?: string;

  /**
   * 基数关系
   */
  cardinality?: DataObjectReferenceCardinality;
};

export type DataObjectJSON = {
  type: DataObjectTypeName.JSON;
};

/**
 * 所有支持的数据类型
 */
export type DataObjectTypeDSL =
  | DataObjectBoolean
  | DataObjectDate
  | DataObjectDateTime
  | DataObjectTimestamp
  | DataObjectInteger
  | DataObjectDecimal
  | DataObjectLong
  | DataObjectDouble
  | DataObjectFloat
  | DataObjectString
  | DataObjectText
  | DataObjectLongText
  | DataObjectJSON
  | DataObjectReference;

/**
 * 数据对象字段
 */
export interface DataObjectPropertyDSL<T extends DataObjectTypeDSL = DataObjectTypeDSL> extends NameDSL {
  /**
   * 表字段名，默认为 name 的 kebab_case 模式
   */
  propertyName?: string;

  /**
   * 字段类型
   */
  type: T;

  /**
   * 是否非空
   */
  notNull?: boolean;

  /**
   * 是否为主键
   */
  primaryKey?: boolean;
}

export enum DataObjectIndexType {
  Normal = 'Normal',
  Unique = 'Unique',
  // PrimaryKey = 'Primary',
  FullText = 'FullText',
}

export enum DataObjectIndexMethod {
  BTREE = 'BTREE',
  HASH = 'HASH',
}

/**
 * 索引描述
 */
export interface DataObjectIndexDSL extends NameDSL {
  /**
   * 索引类型
   */
  type: DataObjectIndexType;

  /**
   * 索引字段
   */
  properties: string[];

  /**
   * 索引方法
   */
  method: DataObjectIndexMethod;
}

/**
 * 数据对象
 */
export interface DataObjectDSL extends NameDSL {
  /**
   * 表名称，默认为 name 的 kebab_case 模式
   */
  tableName?: string;

  /**
   * 字段
   */
  properties: DataObjectPropertyDSL[];

  /**
   * 索引
   */
  indexes: DataObjectIndexDSL[];
}
