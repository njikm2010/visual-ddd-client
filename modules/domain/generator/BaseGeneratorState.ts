import { assert } from '@/lib/utils';
import { NameDSL } from '../domain-design/dsl/dsl';
import { IGeneratorState } from './types';

export class BaseGeneratorState<T extends NameDSL> implements IGeneratorState<T> {
  list: T[] = [];

  results: Map<string, T> = new Map();

  save(uuid: string, result: T) {
    assert(!this.results.has(uuid), 'uuid 重复');

    this.list.push(result);
    this.results.set(uuid, result);
  }

  get(uuid: string): T | undefined {
    return this.results.get(uuid);
  }
}
