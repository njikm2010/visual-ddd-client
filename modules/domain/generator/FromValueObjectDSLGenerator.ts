import { DTODSL, ReferenceTypeDSL, TypeType, ValueObjectDSL } from '../domain-design/dsl/dsl';
import { createDTO, createIDDSL } from '../domain-design/dsl/factory';
import { DTOGeneratorState, IQueryTypeDSLTransformer } from './types';

export class FromValueObjectDSLGenerator {
  private valueObject: ValueObjectDSL;
  private dtoGeneratorState: DTOGeneratorState;
  private queryTypeDSLTransformer: IQueryTypeDSLTransformer;

  constructor(inject: {
    valueObject: ValueObjectDSL;
    dtoGeneratorState: DTOGeneratorState;
    queryTypeDSLTransformer: IQueryTypeDSLTransformer;
  }) {
    const { valueObject, dtoGeneratorState, queryTypeDSLTransformer } = inject;

    this.dtoGeneratorState = dtoGeneratorState;
    this.valueObject = valueObject;
    this.queryTypeDSLTransformer = queryTypeDSLTransformer;
  }

  toDTO(): DTODSL {
    const cache = this.dtoGeneratorState.get(this.valueObject.uuid);
    if (cache) {
      return cache;
    }

    const dto = createDTO();

    dto.name = `${this.valueObject.name}DTO`;
    dto.title = `${this.valueObject.title}数据传输对象`;
    dto.description = `用于${this.valueObject.title}的数据传输对象`;
    dto.properties = [];

    // save cache
    this.dtoGeneratorState.save(this.valueObject.uuid, dto);

    for (const property of this.valueObject.properties) {
      dto.properties.push({
        ...property,
        ...createIDDSL(),
        type: property.type && this.queryTypeDSLTransformer.transformQueryTypeDSL(property.type),
      });
    }

    return dto;
  }

  /**
   * 转换为类型引用
   */
  toDTOTypeDSL(): ReferenceTypeDSL {
    const dto = this.toDTO();

    return {
      type: TypeType.Reference,
      referenceId: dto.uuid,
      name: dto.name,
    };
  }
}
