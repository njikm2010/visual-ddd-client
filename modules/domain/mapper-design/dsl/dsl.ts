import { IDDSL, NameDSL, ReferenceDSL } from '@/modules/domain/domain-design/dsl/dsl';

export type { NameDSL };

export enum ObjectReferenceSource {
  /**
   * 领域对象
   */
  Domain = 'domain',

  /**
   * 结构化对象
   */
  Struct = 'struct',

  /**
   * 数据对象
   */
  Data = 'data',
}

export interface ObjectReferenceDSL extends ReferenceDSL {
  /**
   * 对象的来源，ObjectSelect 依赖它进行回显
   */
  source?: ObjectReferenceSource;
}

export interface FieldMapperDSL extends IDDSL {
  /**
   * 来源字段 uuid, 必填
   */
  source?: string;

  /**
   * 目标字段 uuid，必填
   */
  target?: string;
}

/**
 * 对象结构映射
 */
export interface MapperObjectDSL extends NameDSL {
  /**
   * 来源对象, 领域对象或DTO
   * 必填
   */
  source?: ObjectReferenceDSL;

  /**
   * 目标对象, 数据模型
   * 必填
   */
  target?: ObjectReferenceDSL;

  /**
   * 字段映射
   */
  mappers: FieldMapperDSL[];
}
