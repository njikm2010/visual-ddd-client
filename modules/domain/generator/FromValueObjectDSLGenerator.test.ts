import * as mockedUuid from 'uuid';
import { DTODSL } from '../domain-design/dsl/dsl';
import {
  createBaseType,
  createContainerType,
  createProperty,
  createReferenceType,
  createValueObject,
} from '../domain-design/dsl/factory';

import { BaseGeneratorState } from './BaseGeneratorState';
import { FromValueObjectDSLGenerator } from './FromValueObjectDSLGenerator';

jest.mock('uuid', () => {
  let count = 0;
  return {
    __esModule: true,
    reset() {
      count = 0;
    },
    v4() {
      return `uuid-${count++}`;
    },
  };
});

const mockedUuidModule = mockedUuid as unknown as jest.Mocked<{ reset(): void }>;

describe('FromValueObjectDSLGenerator', () => {
  beforeEach(() => {
    mockedUuidModule.reset();
  });

  it('toDTO baseInfo', () => {
    const valueObject = createValueObject();
    const state = new BaseGeneratorState<DTODSL>();

    const generator = new FromValueObjectDSLGenerator({
      valueObject,
      dtoGeneratorState: state,
      queryTypeDSLTransformer: {
        transformQueryTypeDSL(type) {
          return type;
        },
      },
    });

    const dto = generator.toDTO();

    expect(dto).toMatchObject({
      name: 'UntitledDTO',
      title: '未命名数据传输对象',
      description: '用于未命名的数据传输对象',
    });

    expect(generator.toDTOTypeDSL()).toEqual(createReferenceType(dto.uuid, dto.name));
    expect(state.list).toEqual([dto]);
  });

  it('toDTO properties', () => {
    const state = new BaseGeneratorState<DTODSL>();
    const valueObject = createValueObject();

    valueObject.properties = [];

    // 基础类型
    const baseTypeProperty = createProperty();
    baseTypeProperty.type = createBaseType('String');
    baseTypeProperty.name = 'baseTypeProperty';
    baseTypeProperty.title = '基础类型属性';
    baseTypeProperty.description = '基础类型属性的描述';

    // 容器类型
    const containerTypeProperty = createProperty();
    containerTypeProperty.type = createContainerType('List');
    containerTypeProperty.type.params.item = createBaseType('Boolean');
    containerTypeProperty.name = 'containerTypeProperty';

    // 引用类型 enum
    const referenceTypePropertyForEnum = createProperty();
    referenceTypePropertyForEnum.type = createReferenceType('enum', 'enumName');
    referenceTypePropertyForEnum.name = 'referenceTypePropertyForEnum';

    // 引用类型 valueObject
    const referenceTypePropertyForValueObject = createProperty();
    referenceTypePropertyForValueObject.type = createReferenceType('valueObject', 'valueObjectName');
    referenceTypePropertyForValueObject.name = 'referenceTypePropertyForValueObject';

    valueObject.properties.push(
      baseTypeProperty,
      containerTypeProperty,
      referenceTypePropertyForEnum,
      referenceTypePropertyForValueObject
    );

    const generator = new FromValueObjectDSLGenerator({
      valueObject,
      dtoGeneratorState: state,

      queryTypeDSLTransformer: {
        transformQueryTypeDSL(type) {
          return type;
        },
      },
    });

    const dto = generator.toDTO();

    expect(dto).toMatchSnapshot();
  });
});
