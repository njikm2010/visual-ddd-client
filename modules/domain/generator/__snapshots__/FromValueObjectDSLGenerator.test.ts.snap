// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromValueObjectDSLGenerator toDTO properties 1`] = `
{
  "abstract": false,
  "classMethods": [],
  "classProperties": [],
  "description": "用于未命名的数据传输对象",
  "extends": undefined,
  "implements": [],
  "meta": [],
  "methods": [],
  "name": "UntitledDTO",
  "properties": [
    {
      "access": "public",
      "description": "基础类型属性的描述",
      "meta": [],
      "name": "baseTypeProperty",
      "title": "基础类型属性",
      "type": {
        "name": "String",
        "type": "base",
      },
      "uuid": "uuid-10",
    },
    {
      "access": "public",
      "description": "",
      "meta": [],
      "name": "containerTypeProperty",
      "title": "未命名",
      "type": {
        "name": "List",
        "params": {
          "item": {
            "name": "Boolean",
            "type": "base",
          },
        },
        "type": "container",
      },
      "uuid": "uuid-11",
    },
    {
      "access": "public",
      "description": "",
      "meta": [],
      "name": "referenceTypePropertyForEnum",
      "title": "未命名",
      "type": {
        "name": "enumName",
        "referenceId": "enum",
        "type": "reference",
      },
      "uuid": "uuid-12",
    },
    {
      "access": "public",
      "description": "",
      "meta": [],
      "name": "referenceTypePropertyForValueObject",
      "title": "未命名",
      "type": {
        "name": "valueObjectName",
        "referenceId": "valueObject",
        "type": "reference",
      },
      "uuid": "uuid-13",
    },
  ],
  "title": "未命名数据传输对象",
  "uuid": "uuid-7",
}
`;
